File: app/src/main/java/com/storehop/app/ui/statistics/StatisticsViewModel.kt
package com.storehop.app.ui.statistics

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.storehop.app.data.dao.CategoryPurchaseCount
import com.storehop.app.data.dao.DayCount
import com.storehop.app.data.dao.DayOfWeekCount
import com.storehop.app.data.dao.ItemPurchaseCount
import com.storehop.app.data.dao.StorePurchaseCount
import com.storehop.app.data.repository.CategoryRepository
import com.storehop.app.data.repository.ItemRepository
import com.storehop.app.data.repository.PurchaseHistoryRepository
import com.storehop.app.data.repository.StoreRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import java.time.Clock
import java.time.Duration
import javax.inject.Inject

@HiltViewModel
class StatisticsViewModel @Inject constructor(
    purchases: PurchaseHistoryRepository,
    itemRepository: ItemRepository,
    storeRepository: StoreRepository,
    categoryRepository: CategoryRepository,
    private val clock: Clock,
) : ViewModel() {

    val state: StateFlow<StatisticsUiState> = run {
        val now = clock.millis()
        val thirtyDaysAgo = now - Duration.ofDays(30).toMillis()
        val sevenDaysAgo = now - Duration.ofDays(7).toMillis()
        // Trend chart: last 12 weeks of daily activity. 84-day window keeps
        // the chart legible on a phone screen — long enough to spot a pattern,
        // short enough that low-volume users don't see one tall bar lost in
        // empty days.
        val trendStart = now - Duration.ofDays(TREND_WINDOW_DAYS.toLong()).toMillis()
        val staleCutoff = now - Duration.ofDays(STALE_THRESHOLD_DAYS.toLong()).toMillis()

        // 11 source flows. combine()'s typed overloads cap at 5 arguments;
        // for >5 we spread an Array<Flow<Any?>> through the vararg overload
        // (Flow is covariant on T, so the upcasts are implicit).
        val sources: Array<kotlinx.coroutines.flow.Flow<Any?>> = arrayOf(
            purchases.observeTotalCount(),
            purchases.observeCountSince(thirtyDaysAgo),
            purchases.observeCountSince(sevenDaysAgo),
            purchases.observePurchasesPerDay(trendStart),
            purchases.observePurchasesByDayOfWeek(),
            purchases.observeTopItems(TOP_ITEMS_LIMIT),
            purchases.observePurchasesByStore(),
            purchases.observePurchasesByCategory(),
            itemRepository.observeAll(),
            storeRepository.observeAll(includeArchived = true),
            categoryRepository.observeAll(includeArchived = true),
        )
        combine(*sources) { values ->
            @Suppress("UNCHECKED_CAST")
            buildState(
                totalCount = values[0] as Int,
                last30 = values[1] as Int,
                last7 = values[2] as Int,
                perDay = values[3] as List<DayCount>,
                byDow = values[4] as List<DayOfWeekCount>,
                topItemsRaw = values[5] as List<ItemPurchaseCount>,
                byStoreRaw = values[6] as List<StorePurchaseCount>,
                byCategoryRaw = values[7] as List<CategoryPurchaseCount>,
                items = values[8] as List<com.storehop.app.data.db.relations.ItemWithCategoryAndStores>,
                stores = values[9] as List<com.storehop.app.data.entity.Store>,
                categories = values[10] as List<com.storehop.app.data.entity.Category>,
                staleCutoff = staleCutoff,
                now = now,
            )
        }.stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5_000L),
            initialValue = StatisticsUiState.Loading,
        )
    }

    private fun buildState(
        totalCount: Int,
        last30: Int,
        last7: Int,
        perDay: List<DayCount>,
        byDow: List<DayOfWeekCount>,
        topItemsRaw: List<ItemPurchaseCount>,
        byStoreRaw: List<StorePurchaseCount>,
        byCategoryRaw: List<CategoryPurchaseCount>,
        items: List<com.storehop.app.data.db.relations.ItemWithCategoryAndStores>,
        stores: List<com.storehop.app.data.entity.Store>,
        categories: List<com.storehop.app.data.entity.Category>,
        staleCutoff: Long,
        now: Long,
    ): StatisticsUiState {
        // Totals across the library: useful even when the user has no
        // purchase history yet (so the screen still reads as informative).
        val totalItems = items.size
        val stapleItems = items.count { it.item.isStaple }
        val priorityItems = items.count { it.item.isPriority }

        if (totalCount == 0 && totalItems == 0) return StatisticsUiState.Empty

        val itemNamesById = items.associate { it.item.id to it.item.name }
        val storeNamesById = stores.associate { it.id to it.name }
        val categoryNamesById = categories.associate { it.id to it.name }

        val topItems = topItemsRaw.mapNotNull { row ->
            val name = itemNamesById[row.itemId] ?: return@mapNotNull null
            NamedCount(name = name, count = row.count)
        }

        val purchasesByStore = byStoreRaw.mapNotNull { row ->
            val name = storeNamesById[row.storeId] ?: return@mapNotNull null
            NamedCount(name = name, count = row.count)
        }

        val purchasesByCategory = byCategoryRaw.map { row ->
            val name = if (row.categoryId.isBlank()) {
                null // surfaced as "Uncategorised" via a localised fallback in UI
            } else {
                categoryNamesById[row.categoryId]
            }
            NamedCount(name = name ?: "", count = row.count)
        }

        val staleItems = items
            .asSequence()
            .filter { row ->
                val last = row.item.lastPurchasedAt ?: return@filter false
                last < staleCutoff && !row.item.isNeeded
            }
            .map { row ->
                val last = row.item.lastPurchasedAt ?: 0L
                val days = ((now - last) / DAY_MILLIS).toInt()
                StaleItemRow(name = row.item.name, daysSinceLastPurchase = days)
            }
            .sortedByDescending { it.daysSinceLastPurchase }
            .take(STALE_ITEMS_LIMIT)
            .toList()

        val mostShoppedStore = purchasesByStore.firstOrNull()

        val mostActiveDayOfWeek = byDow.maxByOrNull { it.count }?.dayOfWeek

        return StatisticsUiState.Ready(
            totalPurchases = totalCount,
            purchasesLast30Days = last30,
            purchasesLast7Days = last7,
            purchasesPerDay = perDay,
            mostActiveDayOfWeek = mostActiveDayOfWeek,
            totalItems = totalItems,
            stapleItems = stapleItems,
            priorityItems = priorityItems,
            topItems = topItems,
            staleItems = staleItems,
            totalStores = stores.count { !it.isArchived && it.deletedAt == null },
            mostShoppedStore = mostShoppedStore,
            purchasesByStore = purchasesByStore,
            totalCategories = categories.size,
            purchasesByCategory = purchasesByCategory,
        )
    }

    companion object {
        const val TREND_WINDOW_DAYS = 84
        const val STALE_THRESHOLD_DAYS = 60
        const val TOP_ITEMS_LIMIT = 10
        const val STALE_ITEMS_LIMIT = 20
        private const val DAY_MILLIS = 24L * 60L * 60L * 1000L
    }
}

sealed interface StatisticsUiState {
    data object Loading : StatisticsUiState
    data object Empty : StatisticsUiState
    data class Ready(
        val totalPurchases: Int,
        val purchasesLast30Days: Int,
        val purchasesLast7Days: Int,
        val purchasesPerDay: List<DayCount>,
        /** 0 = Sunday … 6 = Saturday, null when there are no purchases yet. */
        val mostActiveDayOfWeek: Int?,
        val totalItems: Int,
        val stapleItems: Int,
        val priorityItems: Int,
        val topItems: List<NamedCount>,
        val staleItems: List<StaleItemRow>,
        val totalStores: Int,
        val mostShoppedStore: NamedCount?,
        val purchasesByStore: List<NamedCount>,
        val totalCategories: Int,
        val purchasesByCategory: List<NamedCount>,
    ) : StatisticsUiState
}

data class NamedCount(val name: String, val count: Int)
data class StaleItemRow(val name: String, val daysSinceLastPurchase: Int)

File: ios/Storehop/UI/Statistics/StatisticsViewModel.swift
import Foundation
import Observation

/// Mirror of Android's `StatisticsViewModel` — aggregates purchase history
/// + library counts into a single `StatisticsUiState` for the Settings →
/// Statistics screen.
///
/// The Android VM combines 11 flows in one `combine` call; SwiftUI doesn't
/// have a direct equivalent, so we maintain a stash of latest values per
/// source and re-derive the UI state whenever any source changes.
@Observable
@MainActor
final class StatisticsViewModel {
    static let trendWindowDays = 84
    static let staleThresholdDays = 60
    static let topItemsLimit = 10
    static let staleItemsLimit = 20
    private static let dayMillis: Int64 = 24 * 60 * 60 * 1000

    var state: StatisticsUiState = .loading

    // Latest values from each source. Whenever any of these change,
    // `recomputeState()` is called to rebuild `state`.
    private var totalCount: Int = 0
    private var last30: Int = 0
    private var last7: Int = 0
    private var perDay: [DayCount] = []
    private var byDow: [DayOfWeekCount] = []
    private var topItemsRaw: [ItemPurchaseCount] = []
    private var byStoreRaw: [StorePurchaseCount] = []
    private var byCategoryRaw: [CategoryPurchaseCount] = []
    private var items: [ItemWithCategoryAndStores] = []
    private var stores: [Store] = []
    private var categories: [Category] = []

    private let now: Int64
    private let thirtyDaysAgo: Int64
    private let sevenDaysAgo: Int64
    private let trendStart: Int64
    private let staleCutoff: Int64

    private let purchases: PurchaseHistoryRepository
    private let itemRepository: ItemRepository
    private let storeRepository: StoreRepository
    private let categoryRepository: CategoryRepository
    private let session: any UserSessionProvider
    private let totalBinder = SessionBinder()
    private let last30Binder = SessionBinder()
    private let last7Binder = SessionBinder()
    private let perDayBinder = SessionBinder()
    private let byDowBinder = SessionBinder()
    private let topItemsBinder = SessionBinder()
    private let byStoreBinder = SessionBinder()
    private let byCategoryBinder = SessionBinder()
    private let itemsBinder = SessionBinder()
    private let storesBinder = SessionBinder()
    private let categoriesBinder = SessionBinder()

    init(
        purchases: PurchaseHistoryRepository,
        itemRepository: ItemRepository,
        storeRepository: StoreRepository,
        categoryRepository: CategoryRepository,
        session: any UserSessionProvider,
        clock: any Clock
    ) {
        self.purchases = purchases
        self.itemRepository = itemRepository
        self.storeRepository = storeRepository
        self.categoryRepository = categoryRepository
        self.session = session

        let now = clock.nowMs()
        self.now = now
        self.thirtyDaysAgo = now - 30 * Self.dayMillis
        self.sevenDaysAgo = now - 7 * Self.dayMillis
        self.trendStart = now - Int64(Self.trendWindowDays) * Self.dayMillis
        self.staleCutoff = now - Int64(Self.staleThresholdDays) * Self.dayMillis
    }

    func bind() {
        let thirtyDaysAgo = self.thirtyDaysAgo
        let sevenDaysAgo = self.sevenDaysAgo
        let trendStart = self.trendStart

        totalBinder.bind(session: session, emptyValue: 0) { [purchases] uid in
            purchases.observeTotalCount(userId: uid)
        } onValue: { [weak self] v in
            self?.totalCount = v
            self?.recomputeState()
        }
        last30Binder.bind(session: session, emptyValue: 0) { [purchases] uid in
            purchases.observeCountSince(userId: uid, sinceMillis: thirtyDaysAgo)
        } onValue: { [weak self] v in
            self?.last30 = v
            self?.recomputeState()
        }
        last7Binder.bind(session: session, emptyValue: 0) { [purchases] uid in
            purchases.observeCountSince(userId: uid, sinceMillis: sevenDaysAgo)
        } onValue: { [weak self] v in
            self?.last7 = v
            self?.recomputeState()
        }
        perDayBinder.bind(session: session, emptyValue: [DayCount]()) { [purchases] uid in
            purchases.observePurchasesPerDay(userId: uid, sinceMillis: trendStart)
        } onValue: { [weak self] v in
            self?.perDay = v
            self?.recomputeState()
        }
        byDowBinder.bind(session: session, emptyValue: [DayOfWeekCount]()) { [purchases] uid in
            purchases.observePurchasesByDayOfWeek(userId: uid)
        } onValue: { [weak self] v in
            self?.byDow = v
            self?.recomputeState()
        }
        topItemsBinder.bind(session: session, emptyValue: [ItemPurchaseCount]()) { [purchases] uid in
            purchases.observeTopItems(userId: uid, limit: Self.topItemsLimit)
        } onValue: { [weak self] v in
            self?.topItemsRaw = v
            self?.recomputeState()
        }
        byStoreBinder.bind(session: session, emptyValue: [StorePurchaseCount]()) { [purchases] uid in
            purchases.observePurchasesByStore(userId: uid)
        } onValue: { [weak self] v in
            self?.byStoreRaw = v
            self?.recomputeState()
        }
        byCategoryBinder.bind(session: session, emptyValue: [CategoryPurchaseCount]()) { [purchases] uid in
            purchases.observePurchasesByCategory(userId: uid)
        } onValue: { [weak self] v in
            self?.byCategoryRaw = v
            self?.recomputeState()
        }
        itemsBinder.bind(session: session, emptyValue: [ItemWithCategoryAndStores]()) { [itemRepository] uid in
            itemRepository.observeAll(userId: uid)
        } onValue: { [weak self] v in
            self?.items = v
            self?.recomputeState()
        }
        storesBinder.bind(session: session, emptyValue: [Store]()) { [storeRepository] uid in
            storeRepository.observeAll(userId: uid, includeArchived: true)
        } onValue: { [weak self] v in
            self?.stores = v
            self?.recomputeState()
        }
        categoriesBinder.bind(session: session, emptyValue: [Category]()) { [categoryRepository] uid in
            categoryRepository.observeAll(userId: uid, includeArchived: true)
        } onValue: { [weak self] v in
            self?.categories = v
            self?.recomputeState()
        }
    }

    func teardown() {
        totalBinder.cancel()
        last30Binder.cancel()
        last7Binder.cancel()
        perDayBinder.cancel()
        byDowBinder.cancel()
        topItemsBinder.cancel()
        byStoreBinder.cancel()
        byCategoryBinder.cancel()
        itemsBinder.cancel()
        storesBinder.cancel()
        categoriesBinder.cancel()
    }

    // MARK: - State derivation

    private func recomputeState() {
        let totalItems = items.count
        let stapleItems = items.filter { $0.item.isStaple }.count
        let priorityItems = items.filter { $0.item.isPriority }.count

        if totalCount == 0 && totalItems == 0 {
            state = .empty
            return
        }

        let itemNamesById = Dictionary(uniqueKeysWithValues: items.map { ($0.item.id, $0.item.name) })
        let storeNamesById = Dictionary(uniqueKeysWithValues: stores.map { ($0.id, $0.name) })
        let categoryNamesById = Dictionary(uniqueKeysWithValues: categories.map { ($0.id, $0.name) })

        let topItems: [NamedCount] = topItemsRaw.compactMap { row in
            guard let name = itemNamesById[row.itemId] else { return nil }
            return NamedCount(name: name, count: row.count)
        }

        let purchasesByStore: [NamedCount] = byStoreRaw.compactMap { row in
            guard let name = storeNamesById[row.storeId] else { return nil }
            return NamedCount(name: name, count: row.count)
        }

        let purchasesByCategory: [NamedCount] = byCategoryRaw.map { row in
            let name: String?
            if row.categoryId.isEmpty {
                name = nil  // surfaced as "Uncategorised" in the UI
            } else {
                name = categoryNamesById[row.categoryId]
            }
            return NamedCount(name: name ?? "", count: row.count)
        }

        let staleItems: [StaleItemRow] = items
            .compactMap { row -> StaleItemRow? in
                guard let last = row.item.lastPurchasedAt,
                      last < staleCutoff,
                      !row.item.isNeeded else { return nil }
                let days = Int((now - last) / Self.dayMillis)
                return StaleItemRow(name: row.item.name, daysSinceLastPurchase: days)
            }
            .sorted { $0.daysSinceLastPurchase > $1.daysSinceLastPurchase }
            .prefix(Self.staleItemsLimit)
            .map { $0 }

        let mostShoppedStore = purchasesByStore.first
        let mostActiveDayOfWeek = byDow.max(by: { $0.count < $1.count })?.dayOfWeek

        state = .ready(
            StatisticsReadyState(
                totalPurchases: totalCount,
                purchasesLast30Days: last30,
                purchasesLast7Days: last7,
                purchasesPerDay: perDay,
                mostActiveDayOfWeek: mostActiveDayOfWeek,
                totalItems: totalItems,
                stapleItems: stapleItems,
                priorityItems: priorityItems,
                topItems: topItems,
                staleItems: staleItems,
                totalStores: stores.filter { !$0.isArchived && $0.deletedAt == nil }.count,
                mostShoppedStore: mostShoppedStore,
                purchasesByStore: purchasesByStore,
                totalCategories: categories.count,
                purchasesByCategory: purchasesByCategory
            )
        )
    }
}

// MARK: - UI state

enum StatisticsUiState: Equatable, Sendable {
    case loading
    case empty
    case ready(StatisticsReadyState)
}

struct StatisticsReadyState: Equatable, Sendable {
    let totalPurchases: Int
    let purchasesLast30Days: Int
    let purchasesLast7Days: Int
    let purchasesPerDay: [DayCount]
    /// 0 = Sunday … 6 = Saturday, nil when there are no purchases yet.
    let mostActiveDayOfWeek: Int?
    let totalItems: Int
    let stapleItems: Int
    let priorityItems: Int
    let topItems: [NamedCount]
    let staleItems: [StaleItemRow]
    let totalStores: Int
    let mostShoppedStore: NamedCount?
    let purchasesByStore: [NamedCount]
    let totalCategories: Int
    let purchasesByCategory: [NamedCount]
}

struct NamedCount: Equatable, Hashable, Identifiable, Sendable {
    let name: String
    let count: Int
    var id: String { name }
}

struct StaleItemRow: Equatable, Hashable, Identifiable, Sendable {
    let name: String
    let daysSinceLastPurchase: Int
    var id: String { name }
}
